evm/src/contract.ts

```diff
@@ -0,0 +1,179 @@
+// Inspired from https://github.com/arcadeum/multi-token-standard/blob/master/src/tests/utils/contract.ts
+// Build ethers.Contract instances from ABI JSON files generated by truffle.
+//
+// adapted this utility from the handy work by the counterfactual team at:
+// https://github.com/counterfactual/monorepo/blob/d9be8524a691c45b6aac1b5e1cf2ff81059203df/packages/contracts/utils/contract.ts
+
+import * as ethers from 'ethers'
+import { makeDebug } from './debug'
+const debug = makeDebug('abstractContract')
+
+interface NetworkMapping {
+  [networkId: number]: { address: string }
+}
+
+interface BuildArtifact {
+  readonly contractName?: string
+  readonly abi: any[]
+  readonly bytecode: string
+  readonly networks?: NetworkMapping
+}
+
+/**
+ * Convenience class for an undeployed contract i.e. only the ABI and bytecode.
+ */
+export class AbstractContract {
+  /**
+   * Load build artifact by name into an abstract contract
+   * @example
+   *  const CountingApp = AbstractContract.fromArtifactName("CountingApp", { StaticCall })
+   * @param artifactName The name of the artifact to load
+   * @param links Optional AbstractContract libraries to link.
+   * @returns Truffle artifact wrapped in an AbstractContract.
+   */
+  public static fromArtifactName(
+    artifactName: string,
+    links?: Record<string, AbstractContract>,
+  ): AbstractContract {
+    // these ABI JSON files are generated by truffle
+    const contract: BuildArtifact = require(`../../build/contracts/${artifactName}.json`)
+    return AbstractContract.fromBuildArtifact(contract, links, artifactName)
+  }
+
+  /**
+   * Wrap build artifact in abstract contract
+   * @param buildArtifact Truffle contract to wrap
+   * @param links Optional AbstractContract libraries to link.
+   * @returns Truffle artifact wrapped in an AbstractContract.
+   */
+  public static fromBuildArtifact(
+    buildArtifact: BuildArtifact,
+    links?: Record<string, AbstractContract>,
+    artifactName = 'UntitledContract',
+  ): AbstractContract {
+    return new AbstractContract(
+      buildArtifact.abi,
+      buildArtifact.bytecode,
+      buildArtifact.networks,
+      links,
+      artifactName,
+    )
+  }
+
+  public static async getNetworkID(wallet: ethers.Wallet): Promise<number> {
+    return wallet.provider.getNetwork().then(n => n.chainId)
+  }
+
+  /**
+   * @param abi ABI of the abstract contract
+   * @param bytecode Binary of the abstract contract
+   * @param networks Network mapping of deployed addresses
+   * @param links
+   * @param contractName
+   */
+  constructor(
+    readonly abi: any[],
+    readonly bytecode: string,
+    readonly networks: NetworkMapping = {},
+    readonly links?: Record<string, AbstractContract>,
+    readonly contractName?: string,
+  ) {}
+
+  public toStatic(): AbstractContract {
```

Comment:
truffle-contract does this by default, right? Wondering if this is worth doing automatically. Why we have to do this?

Reply:
In the current stable ethers.js version, there is no `.call` method on function calls that we can use. So what we have to do is create an instance of a ethers contract where the ABI has been modified to have its functions be marked as `constant`. Otherwise, ethers will treat it as a non-constant function and perform a transaction instead of a message call. v5 (beta) branch of ethers exposes functionality similar to the `.call` method that truffle-contract exposes, but isnt stable yet. This isn't too big of a deal anyway IMO, since contract functions that should be constant should be marked as such.